libraries/AP_HAL_Linux/UARTDriver.cpp

```diff
@@ -508,30 +416,21 @@ bool UARTDriver::_write_pending_bytes(void)
     }
 
     if (n > 0) {
-        uint16_t n1 = _writebuf_size - _writebuf_head;
-        if (n1 >= n) {
-            // do as a single write
-            _write_fd(&_writebuf[_writebuf_head], n);
+        if (_packetise) {
+            // keep as a single UDP packet
+            uint8_t tmpbuf[n];
+            _writebuf.peekbytes(tmpbuf, n);
```

Comment:
it is `peekbytes()` because it can't advance the pointer before it is sure that was written (and how much was written) in the device, which is done in _write_fd that advances the pointer.

Take in consideration that the buffer's names make harder to understand what should be done, this method does not writes in a buffer, it reads from the writebuffer to write in the device file descriptor (that receives a buffer pointer as argument), so to use ::read() we would have an extra copy because we would write on a temporary buffer to be given to device::write() method and would have an extra problem (and a headache) to deal with partial writes, because ::read() would had advanced the buffer pointer (gg data) before we know that we possible failed to write all the data in the device.

Besides ::read() vs ::peekbytes(), I also think that the relationship between those methods can improve (who 'peeked' the bytes should be responsible for advance it). As it is now (and this behaviour was untouched, it worked like that before), ::_write_fd() receives a buffer as argument but it only works for buffers pointing to a segment of the _writebuf since it always advance it. It is a fragile setup and probably unnecessary.

But, again, remember that this PR is about replace the data structure first and do improvements later. Which I think was a good decision do the amount of feedback we already had on this PR.